### PR TITLE
WIP SS4 compat, rewrite to DBField

### PR DESCRIPTION
**Pull request by chillu**
This started out as an innocent "fix tests for 4.x" task: The master branch of this module (3.x release line) has been made 4.x compatible a while ago, but the test config didn't reflect that

But it turns out the field is quite broken with the SS4 ORM:
The field is not a “composite” field, since it only has one database column.
The DBComposite API has changed a bit in SS4, which causes the field to break in weird ways
since it has overloaded so much of the underlying core functionality. For example,
the field never gets written since it doesn’t set MyFieldValue, and
DataObject->prepareManipulationTable() skips MyField because it’s not a database column.
Rather than fixing these quirks, it’s much easier to rewrite the field as a DBText.

TODO:
- Document db column migration path (from MyFieldValue to MyField)
- Change from serialise() to json_encode() for security reasons

@sanderha @wernerkrauss @digitall-it @brasileric Anybody keen to review and help get this over the line?

### CODE REVIEW

**Comment by nyeholt** [comment]
* Change from serialise() to json_encode() for security reasons

Yep, this is probably my biggest desire with the SS4 move :)

**Comment by jsirish** [comment]
closes #46

File: tests/MultiValueFieldTest.php [hunk 1]
```diff
@@ -1,58 +1,86 @@
...
+        $obj = MultiValueFieldTest_DataObject::get()->byID($obj->ID);
+        $this->assertNotNull($obj->MVField);
+        $this->assertEquals(['One', 'Two'], $obj->obj('MVField')->getValues());
+    }
 
-		$this->assertEquals($second, $obj->MVField->getValues());
-	}
+    public function testSetSerialisedStringAsProperty() {
+        $obj = new MultiValueFieldTest_DataObject();
+        $obj->MVField = serialize(['One', 'Two']);
```
Comment (muskie9) [inline tests/MultiValueFieldTest.php]: Update to `json_encode()` if prior comments are correct.

File: tests/MultiValueFieldTest.php [hunk 2]
```diff
@@ -1,58 +1,86 @@
...
+        $obj = MultiValueFieldTest_DataObject::get()->byID($obj->ID);
+        $this->assertNotNull($obj->MVField);
+        $this->assertEquals(['One', 'Two'], $obj->obj('MVField')->getValues());
+    }
 
-		$field->setValue(array(1, 2, 3));
-		$this->assertTrue($field->isChanged());
+    public function testSetSerialisedStringWithSetValue() {
+        $obj = new MultiValueFieldTest_DataObject();
+        $obj->obj('MVField')->setValue(serialize(['One', 'Two']));
```
Comment (muskie9) [inline tests/MultiValueFieldTest.php]: Update to `json_encode()` if prior comments are correct.

**Comment by muskie9** [comment]
@nyeholt can you comment on what you were going for here and how it is reflected in the updated PR?

**Comment by nyeholt** [comment]
@muskie9  - rather than using serialize and unserialize in get/setValue, it should be using json_encode/json_decode. This could be done in a backwards compatible manner, by looking for "a:" as the starting characters of the field value and assuming the old format, otherwise using json_decode.

**Comment by muskie9** [comment]
I've been digging into this a bit more and am wondering if some classes such as `MultiValueListField` would be more appropriate extending other classes like `SelectField`. This would add to the scope of this PR, but may make more sense based on what the fields are trying to accomplish.

**Comment by muskie9** [comment]
After looking through this some more I did a rebase and [pushed some work up to a fork](https://github.com/muskie9/silverstripe-multivaluefield/commit/b4a60a3e926a99efe3c9b60ed4661a0ac4236e8e) not to muddy this PR. I have all tests passing with the exception of testing `isChanged()`. There seems to be quite a difference in how the `setValue()` methods are currently working in the 3 compatible branch to this PR. Have things changed in SS4 that drastically, or is it just an older PR that doesn't have existing logic brought over?

**Comment by nyeholt** [comment]
I believe Ingo's PR changes the field type that MultiValueField inherits from, so the way setValue works is quite different as it's not longer having to handle writing as a composite field would (which is a much simpler/better approach).

**Comment by muskie9** [comment]
Thanks @nyeholt, after digging into this a bit more, and comparing to some things in framework [`DBCurrency`](http://api.silverstripe.org/en/master/source-class-SilverStripe.ORM.FieldType.DBCurrency.html#5-70)... I'm seeing some noticeable differences. I'm going to try to digest what's going on in the module and framework class a bit more.

**Comment by nyeholt** [comment]
FYI - Release 5.0.0 (https://packagist.org/packages/symbiote/silverstripe-multivaluefield) covers this off now; there's a few reasons that it needs to be a composite field (due to how normal text fields don't give you a chance to modify the field value when it's loaded from the DB if it's not a composite field). This also includes switching to json_encode instead of serialise, but in a BC manner.

**Pull request closed without merging**